### PR TITLE
List project suggestions in create-project command

### PR DESCRIPTION
**Pull request by hkdobrev**
Resolves https://github.com/composer/composer/issues/2752

### CODE REVIEW

**Comment by curry684** [comment]
I understand *why* you implemented it like this but I don't like the implementation. What you did now was add a public utility function to the `Installer` for the sole purpose of injecting, to the installer at that point, arbitrary data into it to trick it into rendering more suggestions at the end. I would like this change not to touch the `Installer` class at all,  as it has no real relationship to it.

This also means that the implementation is flawed, as the suggestions are not rendered now if `--no-install` is supplied, which should have no relationship to whether the root package suggestions are listed. Secondly your PR introduces a member variable in a command class which we are trying to avoid as we want to remove logic and state from the commands some day.

Imo a better implementation would be to add an output parameter to `installRootPackage` returning the suggestions, and then plainly rendering them in the calling function after the `if ($noInstall === false)` section. This way they would always be shown regardless of `--no-install` and introduce a lot less invasive changes.

**Comment by hkdobrev** [comment]
@curry684 Thanks for the pointers! I will try to refactor it the next few days.

Initially I've implemented it only in `CreateProjectCommand`, but the problem is it's not so easy to get all of the installed packages in that command, but even when I did so the output of the suggestions is different places. E.g. You first see the suggested packages from the installed dependencies, then the output of `post-install-cmd` hook, then suggestions from the root project package, then output from `post-create-project-cmd`. Also I've copy-pasted some logic from `Installer` in that implementation.

I think your suggested implementation would work, but at the moment I'm not sure the output would be structured in a nice way and suggestions would be presented together. Will look into it.

**Comment by hkdobrev** [comment]
P.S. I was really amazed of the structure of Composer commands, lack of unit tests for major parts of the system which actually keeps the interface untestable and sometimes confusing.

**Comment by curry684** [comment]
>  I'm not sure the output would be structured in a nice way and suggestions would be presented together

They wouldn't, but that's (again, my opinion) better. The depended upon libraries often suggest packages that just enhance their behavior or add additional interfacing, those suggestions are usually ignored. On the other hand a project that *wants* to be installed through `create-project` would suggest packages really improving the project as a whole. I think it makes sense to show those separately underneath the 'softer' suggestions.

**Comment by hkdobrev** [comment]
@curry684 So you are saying it's better the more important suggestions (from the root package) to be shown below. Because it's more likely the developer would see them without scrolling?

**Comment by curry684** [comment]
That would be the intended result yes. But before we get too far - I'm not the boss here, such details may be better left to @Seldaek to speak on hehe.

**Comment by Seldaek** [comment]
I tend to agree with @curry684 about displaying them at the end. 

What would make sense though to clean things up and avoid duplication is to extract the suggest handling into a new class that would just take an IOInterface, InstalledRepositoryInterface, and have methods like addPackage() to gather suggestions then output() to write it out.

**Comment by hkdobrev** [comment]
@Seldaek Updated it with a separate class taking care of suggestions. I hope this is not too much different than what you had in mind. I've started working on tests for the new class, but wanted to see if that's what you meant.

File: src/Composer/Command/CreateProjectCommand.php [hunk 1]
```diff
@@ -318,6 +327,15 @@ protected function installRootPackage(IOInterface $io, Config $config, $packageN
         $im->install(new InstalledFilesystemRepository(new JsonFile('php://memory')), new InstallOperation($package));
         $im->notifyInstalls($io);
 
+        // collect suggestions
+        foreach ($package->getSuggests() as $target => $reason) {
+            $this->suggestedPackagesReporter->addPackage(
+                $package->getPrettyName(),
+                $target,
+                $reason
+            );
```
Comment (Seldaek) [inline src/Composer/Command/CreateProjectCommand.php]: This block could be de-duplicated into the Reporter class if you just pass the $package

**Comment by Seldaek** [comment]
Looking good yes :)

**Comment by hkdobrev** [comment]
@Seldaek updated with unit tests and code coverage.

**Comment by Seldaek** [comment]
Thanks!

**Pull request merged**